[SPARK-5938][SPARK-5443][SQL] Improve JsonRDD performance

NathanHowell opened the pull request:
This patch comprises of a few related pieces of work:
- Schema inference is performed directly on the JSON token stream
- `String => Row` conversion populate Spark SQL structures without intermediate types
- Projection pushdown is implemented via CatalystScan for DataFrame queries
- Support for the legacy parser by setting `spark.sql.json.useJacksonStreamingAPI` to `false`

Performance improvements depend on the schema and queries being executed, but it should be faster across the board. Below are benchmarks using the last.fm Million Song dataset:

```
Command                                            | Baseline | Patched
---------------------------------------------------|----------|--------
import sqlContext.implicits._                      |          |
val df = sqlContext.jsonFile("/tmp/lastfm.json")   |    70.0s |   14.6s
df.count()                                         |    28.8s |    6.2s
df.rdd.count()                                     |    35.3s |   21.5s
df.where($"artist" === "Robert Hood").collect()    |    28.3s |   16.9s
```

To prepare this dataset for benchmarking, follow these steps:

```
# Fetch the datasets from http://labrosa.ee.columbia.edu/millionsong/lastfm
wget http://labrosa.ee.columbia.edu/millionsong/sites/default/files/lastfm/lastfm_test.zip \
     http://labrosa.ee.columbia.edu/millionsong/sites/default/files/lastfm/lastfm_train.zip

# Decompress and combine, pipe through `jq -c` to ensure there is one record per line
unzip -p lastfm_test.zip lastfm_train.zip  | jq -c . > lastfm.json
```

Commit:
Improve JSON parsing and type inference performance

Commit:
Eliminate arrow pattern, replace with pattern matches

Commit:
Enable JsonRDD2 by default, add a flag to switch back to JsonRDD

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
Jenkins, ok to test.

rxin commented:
I won't have time to look at this today, but this is pretty cool.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31399 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31399/consoleFull) for   PR 5801 at commit [`1abf1d6`](https://github.com/apache/spark/commit/1abf1d6010c71cd1cffa97d7564f8fb71eb19f10).

NathanHowell commented:
Looks like it may also resolve [SPARK-5443](https://issues.apache.org/jira/browse/SPARK-5443).

rxin commented:
Can you put both JIRA tickets in the title? It will then automatically linked to both tickets.

NathanHowell commented:
Done.

SparkQA commented:
[Test build #31399 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31399/consoleFull) for   PR 5801 at commit [`1abf1d6`](https://github.com/apache/spark/commit/1abf1d6010c71cd1cffa97d7564f8fb71eb19f10).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31399/
Test PASSed.

NathanHowell commented:
Benchmarked a small-ish real dataset... Runs are with 5 executors (for 5 input splits) with data in HDFS:

step | before | after 
------|----------|--------
`val df = sqlContext.jsonRDD(...)` - schema inference | 37.14s | 18.16s | 
`df.count()` | 125.8s | 25.7s
`df.select("col1").count()` | 96.9s | 26.5s

Not sure why but the new code seems a bit slower when using projection pushdowns. It may be schema dependent or overhead from evaluating the projection expression.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31449 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31449/consoleFull) for   PR 5801 at commit [`55c2f39`](https://github.com/apache/spark/commit/55c2f391aa727db5ebd62716f3219bb13d236fb2).

SparkQA commented:
[Test build #31449 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31449/consoleFull) for   PR 5801 at commit [`55c2f39`](https://github.com/apache/spark/commit/55c2f391aa727db5ebd62716f3219bb13d236fb2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31449/
Test PASSed.

marmbrus commented:
/cc @yhuai

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31526/consoleFull) for   PR 5801 at commit [`67c381a`](https://github.com/apache/spark/commit/67c381a0f00399133c65ea5a77dda482b441a0aa).

SparkQA commented:
[Test build #31526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31526/consoleFull) for   PR 5801 at commit [`67c381a`](https://github.com/apache/spark/commit/67c381a0f00399133c65ea5a77dda482b441a0aa).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31526/
Test PASSed.

NathanHowell commented:
I think it's in a decent state now, if this qualifies for the 1.4.0 merge window I'll make time to work through any remaining issues (if any).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31544/consoleFull) for   PR 5801 at commit [`bd2e929`](https://github.com/apache/spark/commit/bd2e92926592599a2c31513ec8a7c18874f6a70c).

SparkQA commented:
[Test build #31544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31544/consoleFull) for   PR 5801 at commit [`bd2e929`](https://github.com/apache/spark/commit/bd2e92926592599a2c31513ec8a7c18874f6a70c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31544/
Test PASSed.

yhuai commented:
@NathanHowell This is great! Is it possible to add a feature flag to choose what code path we use? By default, we use the this new code path. But, we still keep the option to use the old one in case there is any issue. Then, in 1.5, we can remove the old code path. What do you think?

NathanHowell commented:
@yhuai Fine with me, I'm reworking the patch set now.

Commit:
Add projection pushdown support to JsonRDD/JsonRDD2

NathanHowell commented:
@yhuai The updated patches do not test the old code. Do you have an opinion on the best way to address this? I can duplicate the entire JsonSuite or try to do something a bit better...

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31626 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31626/consoleFull) for   PR 5801 at commit [`ab6ee87`](https://github.com/apache/spark/commit/ab6ee874ebd72237367bb07dc654c56a54f59375).

marmbrus commented:
I'm okay with freezing the old code and not having tests.  I just want a
quick fall back if a regression is found.
On May 1, 2015 4:01 PM, "Nathan Howell" <notifications@github.com> wrote:

> @yhuai <https://github.com/yhuai> The updated patches do not test the old
> code. Do you have an opinion on the best way to address this? I can
> duplicate the entire JsonSuite or try to do something a bit better...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5801#issuecomment-98258602>.
>

Commit:
Point the empty schema inference test at JsonRDD2

NathanHowell commented:
@marmbrus sounds good, I'll leave it as is.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31630/consoleFull) for   PR 5801 at commit [`842846d`](https://github.com/apache/spark/commit/842846d9c812057c96d23d73319c747c6ac83fca).

SparkQA commented:
[Test build #31626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31626/consoleFull) for   PR 5801 at commit [`ab6ee87`](https://github.com/apache/spark/commit/ab6ee874ebd72237367bb07dc654c56a54f59375).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DataFrameStatFunctions(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31626/
Test PASSed.

SparkQA commented:
[Test build #31630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31630/consoleFull) for   PR 5801 at commit [`842846d`](https://github.com/apache/spark/commit/842846d9c812057c96d23d73319c747c6ac83fca).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31630/
Test PASSed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala:
```diff
@@ -0,0 +1,409 @@
+        parser.nextToken()
+        inferField(parser)
+
+      case VALUE_STRING if parser.getTextLength < 1 => NullType
```
Does it mean that we get an empty string? If so, can we keep the `StringType`? Otherwise, I feel we are destroying information.

NathanHowell replied:
Yes, an empty string gets inferred as a NullType. After inference is
complete any remaining NullType fields get converted back to a StringType.
The old code does this and has test coverage for it, but it does seem a bit
odd.
On May 3, 2015 9:03 PM, "Yin Huai" <notifications@github.com> wrote:

> In sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala
> <https://github.com/apache/spark/pull/5801#discussion_r29565168>:
>
> > +    }
> > +  }
> > +
> > +  /**
> > +   * Infer the type of a json document from the parser's token stream
> > +   */
> > +  private def inferField(parser: JsonParser): DataType = {
> > +    import com.fasterxml.jackson.core.JsonToken._
> > +    parser.getCurrentToken match {
> > +      case null | VALUE_NULL => NullType
> > +
> > +      case FIELD_NAME =>
> > +        parser.nextToken()
> > +        inferField(parser)
> > +
> > +      case VALUE_STRING if parser.getTextLength < 1 => NullType
>
> Does it mean that we get an empty string? If so, can we keep the
> StringType? Otherwise, I feel we are destroying information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5801/files#r29565168>.
>

yhuai replied:
OK, I see. It is for those datasets that use a `""` as a null. Since it is for inferring the type, it is same to use `NullType` because we can always get the `StringType` from other records. Actually, can we add a comment at there to explain its purpose?

NathanHowell replied:
Done.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala:
```diff
@@ -0,0 +1,409 @@
+      case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT, DoubleType) =>
+        parser.getDoubleValue
+
+      case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT, DecimalType()) =>
```
Let's add a `TODO` for handling `DecimalType` with the fixed precision and scale.

NathanHowell replied:
Done.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala:
```diff
@@ -0,0 +1,409 @@
+
+      case _ =>
+        parser.skipChildren()
+        null
```
When will we reach here?

NathanHowell replied:
Hopefully never (at least the compiler doesn't think so), probably leftover crud from development. I've removed it.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala:
```diff
@@ -101,32 +103,83 @@ private[sql] class DefaultSource
-    path: String,
-    samplingRatio: Double,
+private[sql] class JSONRelation(
+    baseRDD: => RDD[String],
```
Should we just use `RDD[String]` as the type?

NathanHowell replied:
Yep, this isn't necessary anymore.

yhuai commented:
@NathanHowell It looks good! Just a few minor comments.

Commit:
Add comments regarding null handling and empty strings

Commit:
Remove unused default case in the field parser

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -67,6 +67,8 @@ private[spark] object SQLConf {
 
   val USE_SQL_SERIALIZER2 = "spark.sql.useSerializer2"
 
+  val USE_JSONRDD2 = "spark.sql.json.useJsonRDD2"
```
maybe useJacksonStreamingAPI?

NathanHowell replied:
Done.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31722/consoleFull) for   PR 5801 at commit [`fa0be47`](https://github.com/apache/spark/commit/fa0be47eb486bd96879dba4a9e48aaf738fd64fa).

Commit:
Add comments on lack of support for precision and scale DecimalTypes

Commit:
JSONRelation's baseRDD need not be lazy

SparkQA commented:
[Test build #31722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31722/consoleFull) for   PR 5801 at commit [`fa0be47`](https://github.com/apache/spark/commit/fa0be47eb486bd96879dba4a9e48aaf738fd64fa).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31722/
Test FAILed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD2.scala:
```diff
@@ -0,0 +1,419 @@
+import org.apache.spark.sql.types._
+import org.apache.spark.Logging
+
+private[sql] object JsonRDD2 extends Logging {
```
would it make sense to name this JacksonStreamingAPI?

rxin replied:
and rename the other one JacksonDataBind

Basically I don't see any connection of this one to "RDD".

NathanHowell replied:
Based on earlier feedback, it seems that this class should replace `JsonRDD` after 1.4.0 is branched. I don't really care that much about the name, since it's internal class (unlike the configuration, which is exposed), but it is more consistent with the work going on in `ParquetRelation2`.

rxin replied:
Yup - but it is really confusing to have this named RDD when it is not an RDD at all. When I was reviewing at the code quickly, I was grepping to find class JsonRDD2, and just couldn't find it ...

NathanHowell replied:
Do you want `JSONRelation` renamed to `JacksonDataBind`?

Commit:
Rename `useJsonRDD2` to `useJacksonStreamingAPI`

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31725 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31725/consoleFull) for   PR 5801 at commit [`b31917b`](https://github.com/apache/spark/commit/b31917b2046c431f5ee0e93e89337ddba9f9c15f).

SparkQA commented:
[Test build #31725 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31725/consoleFull) for   PR 5801 at commit [`b31917b`](https://github.com/apache/spark/commit/b31917b2046c431f5ee0e93e89337ddba9f9c15f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31725/
Test FAILed.

Commit:
Wrap long lines

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31727/consoleFull) for   PR 5801 at commit [`fa8234f`](https://github.com/apache/spark/commit/fa8234fd81d6668f16ee904214c34c35fa56af97).

SparkQA commented:
[Test build #31727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31727/consoleFull) for   PR 5801 at commit [`fa8234f`](https://github.com/apache/spark/commit/fa8234fd81d6668f16ee904214c34c35fa56af97).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31727/
Test PASSed.

Commit:
Split up JsonRDD2 into multiple objects

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31772 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31772/consoleFull) for   PR 5801 at commit [`6822712`](https://github.com/apache/spark/commit/68227127254835787470f4c19bbc67fe01efc76e).

NathanHowell commented:
I split up the rather large `JsonRDD2` object into multiple objects. I picked names that match the Jackson class names but I'm happy to rename them if you have something else in mind.

* `JacksonParser: String => Row`
* `JacksonGenerator: Row => String`
* `InferSchema: String => StructType`

rxin commented:
This looks good to me, but I didn't go through the PR in detail. I will let @yhuai do the final check. Thanks a lot. This is going to be amazing for JSON users!

rxin commented:
@NathanHowell In preparation for merging, do you mind updating the pull request description to include the latest perf numbers, and make sure the old description is still valid?

SparkQA commented:
[Test build #31772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31772/consoleFull) for   PR 5801 at commit [`6822712`](https://github.com/apache/spark/commit/68227127254835787470f4c19bbc67fe01efc76e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31772/
Test PASSed.

NathanHowell commented:
@rxin Do you know if there are any semi-official datasets around for benchmarking? I think it would be better to publish reproducible numbers if possible.

rxin commented:
You can try get some tweets and use that. That's what I usually demo on, but unfortunately I don't think it is legal to make collected tweets public.

There are some stuff here: https://www.opensciencedatacloud.org/publicdata/city-of-chicago-public-datasets/

NathanHowell commented:
@rxin thanks, the datasets there are currently not mounted on their rsync endpoint so I found another (last.fm) dataset that is about 2G and timed a few queries.

rxin commented:
Slightly off topic - @NathanHowell do you know if Jackson allows returning UTF-8 encoded strings directly? If it supports that, we can skip string decoding/encoding altogether, since Spark SQL internally now uses UTF-8 encoded bytes for strings.

NathanHowell commented:
@rxin It supports writing a UTF8 encoding byte array, but there doesn't seem to be equivalent support for reads.. best that can be done is converting the current `char[]` buffer and offset/length directly to a `byte[]`, avoiding an alloc/copy to `String`.

see: http://fasterxml.github.io/jackson-core/javadoc/2.3.0/com/fasterxml/jackson/core/base/ParserMinimalBase.html#getTextCharacters()
and http://fasterxml.github.io/jackson-core/javadoc/2.3.0/com/fasterxml/jackson/core/JsonGenerator.html#writeUTF8String(byte[], int, int)

NathanHowell commented:
@yhuai Is there still time to get this in for 1.4.0?

marmbrus commented:
Yeah, given that there is a flag I think we can still include this.

yhuai commented:
@NathanHowell I will take a final check tomorrow. Can you also add the performance number of selecting all columns in the description? You can use `df.rdd.count` as the command to compare two versions.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -160,6 +162,9 @@ private[sql] class SQLConf extends Serializable {
 
   private[spark] def useSqlSerializer2: Boolean = getConf(USE_SQL_SERIALIZER2, "true").toBoolean
 
+  private[spark] def useJacksonStreamingAPI: Boolean =
```
Can you add comment to explain that it is a temporary flag and we will remove the old code path in 1.5?

yhuai commented:
@NathanHowell I played with it. The issue I found is that `insert` does not work well because `baseRDD` is an input parameter of the JSON relation. For example, with the following code, we will have an exception.
```
sql(
      s"""
        |CREATE TEMPORARY TABLE jsonTable (a int, b string)
        |USING org.apache.spark.sql.json.DefaultSource
        |OPTIONS (
        |  path '/tmp/jsonTable'
        |)
      """.stripMargin)
val rdd = sparkContext.parallelize((1 to 10).map(i => s"""{"a":$i, "b":"str${i}"}"""), 5)
jsonRDD(rdd).registerTempTable("jt")
sql(
      s"""
        |INSERT OVERWRITE TABLE jsonTable SELECT a, b FROM jt
      """.stripMargin)

val rdd = sparkContext.parallelize((1 to 10).map(i => s"""{"a":$i, "b":"str${i}"}"""), 1)
jsonRDD(rdd1).registerTempTable("jt1")
sql(
      s"""
        |INSERT OVERWRITE TABLE jsonTable SELECT a, b FROM jt1
      """.stripMargin)
```

The exception is something like
```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 6.0 failed 1 times, most recent failure: Lost task 3.0 in stage 6.0 (TID 31, localhost): java.io.FileNotFoundException: File file:/tmp/testJson/part-00003 does not exist
	at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:520)
	at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:398)
	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:137)
	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:339)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:763)
	at org.apache.hadoop.mapred.LineRecordReader.<init>(LineRecordReader.java:106)
	at org.apache.hadoop.mapred.TextInputFormat.getRecordReader(TextInputFormat.java:67)
	at org.apache.spark.rdd.HadoopRDD$$anon$1.<init>(HadoopRDD.scala:235)
	at org.apache.spark.rdd.HadoopRDD.compute(HadoopRDD.scala:212)
	at org.apache.spark.rdd.HadoopRDD.compute(HadoopRDD.scala:101)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
	at org.apache.spark.scheduler.Task.run(Task.scala:70)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

```

NathanHowell commented:
@yhuai I'll be able to check on this a bit later today.

yhuai commented:
Seems our test cases are not sufficient to catch the problem. Can you also add the following test cases. 

In `InsertSuite`, let's change the `val rdd` defined in `beforeAll` to `val rdd = sparkContext.parallelize((1 to 10).map(i => s"""{"a":$i, "b":"str${i}"}"""), 5)`. Then, let's change the test of [INSERT OVERWRITE a JSONRelation multiple times](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala#L100-120) to
```
test("INSERT OVERWRITE a JSONRelation multiple times") {
  sql(
    s"""
      |INSERT OVERWRITE TABLE jsonTable SELECT a, b FROM jt
    """.stripMargin)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i, s"str$i"))
  )
  
  // Writing the table to less part files.
  val rdd1 = sparkContext.parallelize((1 to 10).map(i => s"""{"a":$i, "b":"str${i}"}"""), 5)
  jsonRDD(rdd1).registerTempTable("jt1")
  sql(
    s"""
    |INSERT OVERWRITE TABLE jsonTable SELECT a, b FROM jt1
    """.stripMargin)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i, s"str$i"))
  )

  // Writing the table to more part files.
  val rdd2 = sparkContext.parallelize((1 to 10).map(i => s"""{"a":$i, "b":"str${i}"}"""), 10)
  jsonRDD(rdd2).registerTempTable("jt2")
  sql(
    s"""
    |INSERT OVERWRITE TABLE jsonTable SELECT a, b FROM jt2
    """.stripMargin)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i, s"str$i"))
  )

  sql(
    s"""
      |INSERT OVERWRITE TABLE jsonTable SELECT a * 10, b FROM jt1
    """.stripMargin)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i * 10, s"str$i"))
  )

  dropTempTable("jt1")
  dropTempTable("jt2")
}
```
Also, add the following in the `InsertSuite`.
```
test("save directly to the path of a JSON table") {
  table("jt").selectExpr("a * 5 as a", "b").save(path.toString, "json", SaveMode.Overwrite)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i * 5, s"str$i"))
  )

  table("jt").save(path.toString, "json", SaveMode.Overwrite)
  checkAnswer(
    sql("SELECT a, b FROM jsonTable"),
    (1 to 10).map(i => Row(i, s"str$i"))
  )
}
```

Commit:
Add comments to the `useJacksonStreamingAPI` config flag

Commit:
Increase coverage of inserts into a JSONRelation

Commit:
Recreate the baseRDD each for each scan operation

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31981/consoleFull) for   PR 5801 at commit [`e91d8c5`](https://github.com/apache/spark/commit/e91d8c52fa2a83b3031444fcce3874ca5a3a7440).

SparkQA commented:
[Test build #31981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31981/consoleFull) for   PR 5801 at commit [`e91d8c5`](https://github.com/apache/spark/commit/e91d8c52fa2a83b3031444fcce3874ca5a3a7440).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31981/
Test PASSed.

NathanHowell commented:
@yhuai I've added the tests and fixed the failures, the change was minor... changed the type of baseRDD back to `=> RDD[String]` and added some comments.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala:
```diff
@@ -101,32 +103,84 @@ private[sql] class DefaultSource
+private[sql] class JSONRelation(
+    // baseRDD needs to be created on scan and not when JSONRelation is
+    // constructed, so we need a function (call by name) instead of a value
+    baseRDD: => RDD[String],
```
can you update the comment to document why it needs to be created on scan?

yhuai replied:
How about we explicitly pass a closure (more reader friendly)?

NathanHowell commented:
@rxin yep, I've updated the comment.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32000/consoleFull) for   PR 5801 at commit [`e1187eb`](https://github.com/apache/spark/commit/e1187eb0e1263fba1fa53f64fad5917649997869).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala:
```diff
@@ -101,32 +103,87 @@ private[sql] class DefaultSource
+    // underlying inputs are modified. To be safe, a call-by-name
+    // value (a function) is used instead of a regular value to
+    // ensure the RDD is recreated on each and every operation.
+    baseRDD: => RDD[String],
```
@NathanHowell Do you think a closure at here will be better (as mentioned in https://github.com/databricks/scala-style-guide#call_by_name)?

NathanHowell replied:
Yes, this is corrected.

SparkQA commented:
[Test build #32000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32000/consoleFull) for   PR 5801 at commit [`e1187eb`](https://github.com/apache/spark/commit/e1187eb0e1263fba1fa53f64fad5917649997869).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32000/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32001/consoleFull) for   PR 5801 at commit [`26fea31`](https://github.com/apache/spark/commit/26fea314ab12657df6c6397f832d35e681d2e27a).

SparkQA commented:
[Test build #32001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32001/consoleFull) for   PR 5801 at commit [`26fea31`](https://github.com/apache/spark/commit/26fea314ab12657df6c6397f832d35e681d2e27a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32001/
Test PASSed.

yhuai commented:
Thank you! LGTM. I am merging it to master and branch 1.4.
